docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -188,6 +193,29 @@ type DockerMachineTemplate struct {
 ```
 _Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_
 
+`platform.NodeSystemInfo` is a struct that contains the architecture and operating system information of the node, to 
+implement in the `util` package of the `cluster-api` project. The defined types and constants are exported for use by the 
+cluster-api providers integrations.
+Its definition would look like this:
```

Comment:
I think we should not put an API struct in the util package. 

On the other side I'm also not sure if we should put it in any of our existing packages of our api groups. Because as soon we create a new apiVersion every infra provider CRD also requires a new apiVersion to bump to our new apiVersion.

So I think it might be a good idea to create a new API package for this. One option could be `api/clusterautoscaler/<version>` (could be fine to directly start with version v1).

In general I would like to see this API versioned, so we have a chance to evolve it over time.

Reply:
Hi @sbueringer, I'm looking into the api package. Does it make more sense to put these structs in the `api/${version:-v1beta1}/` package? Perhaps we can have a `machinetemplate_types.go` file as we do for other APIs.

I'm not sure about `api/clusterautoscaler/<version>` because (a) other types are defined in `api/<version>/...`, and (b) the clusterautoscaler may not be the direct consume of these types as it reads the MachineTemplate as an unstructured object. Providers, though, should use these types to expose the platform information to the cluster autoscaler.

Reply:
> api/<version>/...

Will be going away soon. We'll move to something like this
```
api
  core
  addons
  bootstrap/kubeadm
  controlplane/kubeadm
```

I think we should not put it into the core API package (today api/v1beta1, later api/core/v1beta1) as it is not part of the core CAPI API.

Regarding the package name. What about `api/infrastructure/\<version\>`?

But I would like to hear opinions from other maintainers

Reply:
Just to mention it. Another alternative could be to just define the struct here in this KEP and then infra providers just copy it over into their API types. Would also avoid coupling between core CAPI and infra providers (i.e. if CAPI does an API bump that affects the core CAPI struct, providers are not affected).

This might be okay as the actual dependency is to the cluster-autoscaler code that uses these fields

Reply:
cc @fabriziopandini @chrischdi @JoelSpeed

Reply:
> then infra providers just copy it over into their API types.

Do we expect every provider to implement every single field? If we define the schema here for each possible field, then the providers can copy over only the fields they support, and not have to add fields to their API that they may not actually do anything with.

If they later add a new capability, they just need to update by looking at how the field was defined here

Reply:
I'm definitely fine with the copy approach. Embedding structs is really tricky and probably not worth the coupling

Reply:
Updated.